pkg/csi/service/node.go

```diff
@@ -674,109 +683,272 @@ func (driver *vsphereCSIDriver) NodeGetInfo(
-	if err != nil {
-		if os.IsNotExist(err) {
-			log.Infof("Config file not provided to node daemonset. Assuming non-topology aware cluster.")
-			nodeInfoResponse = &csi.NodeGetInfoResponse{
-				NodeId:            nodeID,
-				MaxVolumesPerNode: maxVolumesPerNode,
+
+	var (
+		accessibleTopology map[string]string
+		err error
+	)
+	if commonco.ContainerOrchestratorUtility.IsFSSEnabled(ctx, common.ImprovedVolumeTopology) {
+		accessibleTopology, err = fetchTopologyLabelsUsingCR(ctx, nodeName, nodeID)
+	} else {
+		var cfg *cnsconfig.Config
+		cfgPath = os.Getenv(cnsconfig.EnvVSphereCSIConfig)
+		if cfgPath == "" {
+			cfgPath = cnsconfig.DefaultCloudConfigPath
+		}
+		cfg, err := cnsconfig.GetCnsconfig(ctx, cfgPath)
+		if err != nil {
+			if os.IsNotExist(err) {
+				log.Infof("Config file not provided to node daemonset. Assuming non-topology aware cluster.")
+				nodeInfoResponse = &csi.NodeGetInfoResponse{
+					NodeId:            nodeID,
+					MaxVolumesPerNode: maxVolumesPerNode,
+				}
+				log.Infof("NodeGetInfo response: %v", nodeInfoResponse)
+				return nodeInfoResponse, nil
 			}
-			log.Infof("NodeGetInfo response: %v", nodeInfoResponse)
-			return nodeInfoResponse, nil
+			return nil, logger.LogNewErrorCodef(log, codes.Internal,
+				"failed to read CNS config. Error: %v", err)
 		}
-		return nil, logger.LogNewErrorCodef(log, codes.Internal,
-			"failed to read cnsconfig. err: %v", err)
+		accessibleTopology, err = fetchTopologyLabelsUsingVCCreds(ctx, nodeID, cfg)
+	}
+	if err != nil {
+		return nil, err
 	}
-	var accessibleTopology map[string]string
+
 	topology := &csi.Topology{}
+	if len(accessibleTopology) > 0 {
+		topology.Segments = accessibleTopology
+	}
+	nodeInfoResponse = &csi.NodeGetInfoResponse{
+		NodeId:             nodeID,
+		MaxVolumesPerNode:  maxVolumesPerNode,
+		AccessibleTopology: topology,
+	}
+	log.Infof("NodeGetInfo response: %v", nodeInfoResponse)
+	return nodeInfoResponse, nil
+}
 
-	if cfg.Labels.Zone != "" && cfg.Labels.Region != "" {
-		log.Infof("Config file provided to node daemonset with zones and regions. Assuming topology aware cluster.")
-		vcenterconfig, err := cnsvsphere.GetVirtualCenterConfig(ctx, cfg)
-		if err != nil {
+func fetchTopologyLabelsUsingCR(ctx context.Context, nodeName, nodeID string) (map[string]string, error) {
+	log := logger.GetLogger(ctx)
+
+	// Get kube config.
+	config, err := k8s.GetKubeConfig(ctx)
+	if err != nil {
+		return nil, logger.LogNewErrorCodef(log, codes.Internal, "failed to fetch kubeconfig. Error: %+v", err)
+	}
+	// Create controller runtime client.
+	controllerRuntimeClient, err := k8s.NewClientForGroup(ctx, config, csinodetopologyv1alpha1.GroupName)
+	if err != nil {
+		return nil, logger.LogNewErrorCodef(log, codes.Internal,
+			"failed to create controllerRuntimeClient. Error: %+v", err)
+	}
+
+	// Fetch node object to set owner ref.
+	nodeObj, err := getNodeObject(ctx, nodeName)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, err.Error())
```

Comment:
How about using `LogNewErrorCodef` here?

Reply:
I do not need to log anything here. The function `getNodeObject` logs all errors before returning.